jobs/fix_crt_channel_unreads/worker.go

```diff
@@ -0,0 +1,210 @@
+	close(w.stopChan)
+	<-w.stoppedChan
+}
+
+func (w *FixCRTChannelUnreadsWorker) doJob(job *model.Job) {
+	if claimed, err := w.jobServer.ClaimJob(job); err != nil {
+		mlog.Warn("Worker experienced an error while trying to claim job",
+			mlog.String("worker", w.name),
+			mlog.String("job_id", job.Id),
+			mlog.String("error", err.Error()))
+		return
+	} else if !claimed {
+		return
+	}
+	if _, err := w.app.Srv().Store.System().GetByName(model.MigrationKeyFixCRTChannelUnreads); err == nil {
+		mlog.Info("Worker: migration already done", mlog.String("worker", w.name), mlog.String("job_id", job.Id))
+		w.setJobSuccess(job)
+		return
+	}
+
+	userID := ""
+	channelID := ""
+	updatedInThisRun := 0
+	fixedBadCM := 0
+	migrationDone := false
+	for {
+		cms, sErr := w.app.Srv().Store.Channel().GetCRTUnfixedChannelMembershipsAfter(userID, channelID, 100)
+		if sErr != nil {
+			if sErr == sql.ErrNoRows {
+				migrationDone = true
+				break
+			}
+			mlog.Warn("Failed to get bad channel memberships",
+				mlog.String("worker", w.name),
+				mlog.String("job_id", job.Id),
+				mlog.String("error", sErr.Error()))
+			continue
+		}
+		lastCM := cms[len(cms)-1]
+		userID = lastCM.UserId
+		channelID = lastCM.ChannelId
+
+		markAsFixed := make([]model.ChannelMember, 0, len(cms))
+
+		for _, cm := range cms {
+			isUnread, err := w.app.Srv().Store.Channel().IsChannelMemberUnread(cm, true)
+			if err != nil {
+				mlog.Warn("Worker experienced an error while trying to get channel unread",
+					mlog.String("worker", w.name),
+					mlog.String("job_id", job.Id),
+					mlog.String("error", err.Error()))
+				continue
+			}
+			if !isUnread {
+				updatedInThisRun++
+				markAsFixed = append(markAsFixed, cm)
+				continue
+			}
+
+			postTypes, err := w.app.Srv().Store.Post().GetUniquePostTypesSince(cm.ChannelId, cm.LastViewedAt)
+			if err != nil {
+				continue
+			}
+			if !containsNormalPost(postTypes) {
+				_, err := w.app.Srv().Store.Channel().UpdateLastViewedAt([]string{cm.ChannelId}, cm.UserId, false)
+				if err != nil {
+					mlog.Warn("Worker experienced an error while trying to mark channel as read",
+						mlog.String("worker", w.name),
+						mlog.String("job_id", job.Id),
+						mlog.String("error", err.Error()))
+					continue
+				}
+				fixedBadCM++
+			}
+			markAsFixed = append(markAsFixed, cm)
+			updatedInThisRun++
+		}
+		err := w.app.Srv().Store.Channel().MarkChannelMembersAsCRTFixed(markAsFixed)
+		if err != nil {
+			mlog.Warn("Worker experienced an error while trying to mark channel memberships as fixed",
+				mlog.String("worker", w.name),
+				mlog.String("job_id", job.Id),
+				mlog.String("error", err.Error()))
+		}
+	}
+
+	if migrationDone {
+		system := model.System{
+			Name:  model.MigrationKeyFixCRTChannelUnreads,
+			Value: "true",
+		}
+
+		if err := w.app.Srv().Store.System().Save(&system); err != nil {
+			mlog.Critical("Failed to mark crt channel unreads migration job as completed.", mlog.Err(err))
```

Comment:
Gentle reminder to check our logging [principles](https://developers.mattermost.com/contribute/server/style-guide/#log-levels) before merging this change.